CRVResponse/src/CrvStepsFromStepPointMCs_module.cc

```diff
@@ -0,0 +1,361 @@
+          {
+	    cout << " CrvSteps with " << spmcIndices[i].second-spmcIndices[i].first+1 << " steps,"
+                 << " barIndex = " << crvStep.barIndex()  << " SimParticle Key = " << crvStep.simParticle()->id()
+                 << " edep = " << crvStep.visibleEDep() << " pathlen = " << crvStep.pathLength()
+                 << " glen = " << sqrt((crvStep.endPos()-crvStep.startPos()).mag2())
+                 << " startTime = " << crvStep.startTime() << " endTime = " << crvStep.endTime() << endl;
+          }
+        }
+      } // end of pair loop
+    } // end of collection loop
+    if(_debug > 0)
+    {
+      cout << "Total number of CrvSteps " << crvSteps->size() << " , StepPointMCs = " << nspmcs << endl;
+    }
+    event.put(move(crvSteps));
+  } // end of produce
+
+  void CrvStepsFromStepPointMCs::fillSteps(SPMCPtrV const& spmcptrv, CRSScintillatorBarIndex const& barIndex,
+                                           ParticleData const* pdata, cet::map_vector_key trackId, 
+                                           unique_ptr<CrvStepCollection> &crvSteps, vector<pair<size_t,size_t> > &spmcIndices)
+  {
+    //the sequence of steps needs to be devided into two (or more) CrvSteps, if the particle leaves the scintillator and returns later
+    //this situtation can be identified by the end process code of Transportation
+    bool newCrvStep(true);
+
+    // variables we accumulate for all the StepPoints in this pair
+    double edep(0.0), pathlen(0.0);
+    // keep track of the first and last step
+    SPMCPtr first;
+    SPMCPtr last;
+    // loop over all the StepPoints
+    for(size_t i=0; i<=spmcptrv.size(); i++)
+    {
+      if(i<spmcptrv.size())
+      {
+        const auto& spmcptr = spmcptrv[i];
+        if(newCrvStep)
+        {
+          newCrvStep=false;
+          edep      =0.0;
+          pathlen   =0.0;
+          first     =spmcptr;
+          spmcIndices.emplace_back(i,i);
+        }
+        edep   +=spmcptr->visibleEDep();
+        pathlen+=spmcptr->stepLength();
+        last    =spmcptr;
+        spmcIndices.back().second=i;
+
+        if(spmcptr->endProcessCode()==ProcessCode::Transportation)
+        {
+          newCrvStep=true;
+        }
+      }
+
+      //analyze the accumulated steps at the end of the sequence or if the track leaves the scintillator
+      if(i==spmcptrv.size() || newCrvStep)
+      {
+        if(first.isNull() || last.isNull())
+          throw cet::exception("SIM")<<"mu2e::CrvStepsFromStepPointMCs: No first or last step" << endl;
+
+        // Define the position at entrance and exit
+        CLHEP::Hep3Vector startPos = first->position();
+        CLHEP::Hep3Vector endPos   = last->postPosition();
+
+        double startTime = first->time();
+        double endTime   = last->time();
+
+        CLHEP::Hep3Vector startMomV = first->momentum();
+        CLHEP::Hep3Vector endMomV   = last->momentum();
+
+        //TODO: would be nice to have an end momentum in the StepPointMC, 
```

Comment:
Do you refer to the scalar change in momentum or also direction change?

Reply:
Downstream modules only need the scalar change to calculate an average value of beta. However, it is possible that we may find a use for the direction change later.